pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -127,83 +93,58 @@ pub fn is_ephemeral_file(filename: &str) -> bool {
-            .map_err(|e| to_io_error(e, "Failed to write ephemeral buf"))?
-        {
-            WriteBufResult::Found(guard) => {
-                write_guard = guard;
-                write_guard.deref_mut()
-            }
-            WriteBufResult::NotFound(guard) => {
-                // Read the page from disk into the buffer
-                // TODO: if we're overwriting the whole page, no need to read it in first
-                write_guard = guard;
-                self.fill_buffer(write_guard.deref_mut(), blkno)?;
-                write_guard.mark_valid();
-
-                // And then fall through to modify it.
-                write_guard.deref_mut()
-            }
-        };
-
-        buf[off..(off + len)].copy_from_slice(&srcbuf[0..len]);
-        write_guard.mark_dirty();
-        Ok(len)
-    }
-}
-
 impl BlobWriter for EphemeralFile {
     fn write_blob(&mut self, srcbuf: &[u8]) -> Result<u64, io::Error> {
         let pos = self.size;
 
+        // TODO: rewrite this to use bytes::BytesMut / a wrapper that
+        // auto-flushes underneath. It'd be much less error-prone.
+
         let mut blknum = (self.size / PAGE_SZ as u64) as u32;
         let mut off = (pos % PAGE_SZ as u64) as usize;
 
-        let mut buf = self.get_buf_for_write(blknum)?;
+        let buf = &mut self.mutable_head;
+        let flush_head = |head: &mut [u8], blknum: u32| {
+            debug_assert_eq!(head.len(), PAGE_SZ);
+            match self.file.write_all_at(head, blknum as u64 * PAGE_SZ as u64) {
+                Ok(_) => {
+                    // Pre-warm the page cache with what we just wrote.
+                    // This isn't necessary for coherency/correctness, but it's how we've always done it.
+                    let cache = page_cache::get();
+                    match cache.read_immutable_buf(self.page_cache_file_id, blknum) {
+                        Err(e) => {
+                            error!("ephemeral_file flush_head failed to get immutable buf to pre-warm page cache: {e:?}");
+                            // fail gracefully, it's not the end of the world if we can't pre-warm the cache here
+                        }
+                        Ok(page_cache::ReadBufResult::Found(_guard)) => {
+                            // This function takes &mut self, so, it shouldn't be possible to reach this point.
+                            unreachable!("we just wrote blknum {blknum} and this function takes &mut self, so, no concurrent read_blk is possible");
+                        }
+                        Ok(page_cache::ReadBufResult::NotFound(mut write_guard)) => {
+                            let buf: &mut [u8] = write_guard.deref_mut();
+                            debug_assert_eq!(buf.len(), PAGE_SZ);
+                            buf.copy_from_slice(head);
+                            write_guard.mark_valid();
+                            // pre-warm successful
+                        }
```

Comment:
All rust tutorials, guides, libraries everywhere always process variants in the order in which enums are introduced. No need for adhoc better style :tm: here:
```suggestion
                        Ok(page_cache::ReadBufResult::Found(_guard)) => {
                            // This function takes &mut self, so, it shouldn't be possible to reach this point.
                            unreachable!("we just wrote blknum {blknum} and this function takes &mut self, so, no concurrent read_blk is possible");
                        }
                        Ok(page_cache::ReadBufResult::NotFound(mut write_guard)) => {
                            let buf: &mut [u8] = write_guard.deref_mut();
                            debug_assert_eq!(buf.len(), PAGE_SZ);
                            buf.copy_from_slice(head);
                            write_guard.mark_valid();
                            // pre-warm successful
                        }
                        Err(e) => {
                            error!("ephemeral_file flush_head failed to get immutable buf to pre-warm page cache: {e:?}");
                            // fail gracefully, it's not the end of the world if we can't pre-warm the cache here
                        }
```

Reply:
Resolved since I rebase-merged this PR on top of #5004

Reply:
Hmm #5004 did not change this.

Reply:
Right, https://github.com/neondatabase/neon/pull/4994/commits/b56edb4ee353fca2b8e026110e6b3272f58b9676 did